### PR TITLE
Add Namespace Field.

### PR DESCRIPTION
**Pull request by davinchia**
## What
Add namespace field to the Airbyte Stream in preparation to propagate a source defined namespace to the Destination.

This namespace field is then consumed as the destination schema the table is written to.

This only applies to JDBC destinations.

This is Steps 1 - 4 of this [tech spec](https://docs.google.com/document/d/1qFk4YqnwxE4MCGeJ9M2scGOYej6JnDy9A0zbICP_zjI/edit#heading=h.ab3zcl6xt7hb).

Some minor refactoring and commenting as I go.

## How
See the above.

## Pre-merge Checklist
[x] Run integration test.

## Recommended reading order
1. `airbyte_protocol.yaml` (main change)
2. `JdbcBufferedConsumerFactory` (propagate change to JDBC destinations)
3. `BigQueryDestination` (propagate change to BigQuery)
4. `RedshiftCopyDestination` and `RedshiftCopier` (propagate change to Redshift)
5. `TestDestination` and various `YDestination` classes (understand how this change tests)
6. `spec.json` (UI changes)
7. `en.json` (UI changes)

### CODE REVIEW

File: airbyte-integrations/connectors/destination-jdbc/src/main/java/io/airbyte/integrations/destination/jdbc/WriteConfig.java [hunk 1]
```diff
@@ -22,21 +22,24 @@
...
 import io.airbyte.protocol.models.DestinationSyncMode;
 
+/**
+ * Write configuration POJO for all destinations extending {@link AbstractJdbcDestination}.
+ */
 public class WriteConfig {
 
   private final String streamName;
-  private final String outputNamespaceName;
+  private final String outputSchemaName;
```
Comment (davinchia) [inline airbyte-integrations/connectors/destination-jdbc/src/main/java/io/airbyte/integrations/destination/jdbc/WriteConfig.java:35]: Since namespace is something that doesn't appear in the destination concept space now, I felt it was clearer to rename this field.

**Comment by davinchia** [comment]
@cgardens want to take a quick sanity check? The destination side implementation is very simple - mostly struggling with testing this on the non-locally available tests. I'm pretty confident; just 2 tests cases and this should be good to go.

File: airbyte-integrations/bases/standard-destination-test/src/main/java/io/airbyte/integrations/standardtest/destination/DataArgumentsProvider.java [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ */
+
+package io.airbyte.integrations.standardtest.destination;
```
Comment (davinchia) [inline airbyte-integrations/bases/standard-destination-test/src/main/java/io/airbyte/integrations/standardtest/destination/DataArgumentsProvider.java:25]: Broke this out since the `TestDestination` class was getting quite large.

File: airbyte-integrations/bases/standard-destination-test/src/main/java/io/airbyte/integrations/standardtest/destination/DataArgumentsProvider.java [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+import org.junit.jupiter.params.provider.ArgumentsProvider;
+
+/**
+ * Class encapsulating all arguments required for Standard Destination Tests.
+ *
+ * All files defined here can be found in src/main/resources of this package.
+ */
+public class DataArgumentsProvider implements ArgumentsProvider {
+
+  public static final CatalogMessageTestConfigPair EXCHANGE_RATE_CONFIG =
```
Comment (davinchia) [inline airbyte-integrations/bases/standard-destination-test/src/main/java/io/airbyte/integrations/standardtest/destination/DataArgumentsProvider.java:39]: also allows us to centralise naming and references to the various test configs here instead of scattered manually typed strings.

File: airbyte-integrations/connectors/destination-jdbc/src/test/java/io/airbyte/integrations/destination/jdbc/JdbcDestinationTest.java [hunk 4]
```diff
@@ -1,335 +0,0 @@
-/*
```
Comment (davinchia) [inline airbyte-integrations/connectors/destination-jdbc/src/test/java/io/airbyte/integrations/destination/jdbc/JdbcDestinationTest.java:1]: I decided that the test here were of very little use since:
1) they are almost an exact duplicate of the JDBC Integration tests in terms of what is being tested.
2) they serve no real value since the JDBC integration test can be run locally without additional credentials.
3) the main value this class brings is letting us run tests without building the docker image (the integration tests require doing so). however I feel this benefit is not worth the additional maintenance cost.

File: airbyte-integrations/bases/standard-destination-test/src/main/java/io/airbyte/integrations/standardtest/destination/TestDestination.java [hunk 5]
```diff
@@ -127,20 +122,12 @@
    *
```
Comment (davinchia) [inline airbyte-integrations/bases/standard-destination-test/src/main/java/io/airbyte/integrations/standardtest/destination/TestDestination.java:122]: There are many changes in this file since I did a fair amount of refactoring here for readability.

Main changes are adding a `namespace` field to the `retrieveRecords` and `retrieveNormalizedRecords` functions.

I rearrange the test methods so it's
1) Get
2) Check(s)
3) Syncs
4) Incremental
5) Normalisation
6) Namespace

And pushed all the helper functions down.

File: airbyte-protocol/models/src/main/resources/airbyte_protocol/airbyte_protocol.yaml [hunk 6]
```diff
@@ -151,6 +151,9 @@ definitions:
           type: array
           items:
             type: string
+      namespace:
+        description: Optional Source-defined namespace. Currently only used by JDBC destinations to determine what schema to write to. Airbyte streams from the same sources should have the same namespace.
```
Comment (sherifnada) [inline airbyte-protocol/models/src/main/resources/airbyte_protocol/airbyte_protocol.yaml:155]: Based on this description it feels like this concept should not live in the stream. If this information only makes sense to destinations then shouldn't it be defined in the connection or destination config?

**Review by sherifnada — CHANGES_REQUESTED** [review]
One blocking question

File: airbyte-protocol/models/src/main/resources/airbyte_protocol/airbyte_protocol.yaml (hunk 6 above)
Comment (sherifnada) [reply airbyte-protocol/models/src/main/resources/airbyte_protocol/airbyte_protocol.yaml:155]: nvm this makes sense. We're just propagating the schema. I think I got a little thrown off by the description into thinking this is a destination feature that lives in the source.

**Review by sherifnada — COMMENTED** [review]
Worked out my question - no longer blocking

**Comment by davinchia** [comment]
@sherifnada my tech spec clear? how did you work it out?

**Comment by davinchia** [comment]
I refactored parts of `TestDestination`, so going to test everything once more.

File: airbyte-integrations/connectors/destination-bigquery/src/main/resources/spec.json [hunk 7]
```diff
@@ -16,8 +16,8 @@
       },
       "dataset_id": {
         "type": "string",
-        "description": "The BigQuery dataset id that will house replicated tables.",
-        "title": "Dataset ID"
+        "description": "Default BigQuery Dataset ID tables are replicated to if the source does not specify a schema.",
```
Comment (davinchia) [inline airbyte-integrations/connectors/destination-bigquery/src/main/resources/spec.json:19]: Result: 
Screen Shot 2021-03-27 at 1 38 45 AM

File: airbyte-integrations/connectors/destination-postgres/src/main/resources/spec.json [hunk 8]
```diff
@@ -32,8 +32,8 @@
         "order": 2
       },
       "schema": {
-        "title": "Schema",
-        "description": "Unless specifically configured, the usual value for this field is \"public\".",
+        "title": "Default Schema",
```
Comment (davinchia) [inline airbyte-integrations/connectors/destination-postgres/src/main/resources/spec.json:35]: Result:
Screen Shot 2021-04-04 at 6 10 56 PM

File: airbyte-integrations/connectors/destination-redshift/src/main/resources/spec.json [hunk 9]
```diff
@@ -11,34 +11,40 @@
     "properties": {
       "host": {
         "description": "Host Endpoint of the Redshift Cluster (must include the cluster-id, region and end with .redshift.amazonaws.com)",
-        "type": "string"
```
Comment (davinchia) [inline airbyte-integrations/connectors/destination-redshift/src/main/resources/spec.json:14]: Result: 
Screen Shot 2021-04-04 at 6 11 31 PM

File: airbyte-integrations/connectors/destination-snowflake/src/main/resources/spec.json [hunk 10]
```diff
@@ -21,37 +21,44 @@
       "host": {
         "description": "Host domain of the snowflake instance (must include the account, region, cloud environment, and end with snowflakecomputing.com).",
         "examples": ["accountname.us-east-2.aws.snowflakecomputing.com"],
-        "type": "string"
```
Comment (davinchia) [inline airbyte-integrations/connectors/destination-snowflake/src/main/resources/spec.json:24]: Result:
image

File: airbyte-webapp/src/locales/en.json [hunk 11]
```diff
@@ -76,8 +76,8 @@
   "form.edit": "Edit",
   "form.done": "Done",
   "form.examples": "(e.g. {examples})",
-  "form.prefix": "Namespace Prefix",
-  "form.prefix.message": "Where to replicate your data to in your destination",
+  "form.prefix": "Table Prefix",
```
Comment (davinchia) [inline airbyte-webapp/src/locales/en.json:79]: Result: 
image

**Comment by davinchia** [comment]
@cgardens @sherifnada @ChristopheDuong besides publishing the new docker images, the destination side of enabling namespaces is done, so please take a look. Will follow this up with a source side PR to enable this for JDBC sources.

File: airbyte-integrations/connectors/destination-bigquery/src/main/resources/spec.json (hunk 7 above)
Comment (cgardens) [reply airbyte-integrations/connectors/destination-bigquery/src/main/resources/spec.json:19]: i think this might be the wrong screenshot? i think this is the one for redshift?

File: airbyte-integrations/connectors/destination-postgres/src/main/resources/spec.json (hunk 8 above)
Comment (cgardens) [reply airbyte-integrations/connectors/destination-postgres/src/main/resources/spec.json:35]: >  source does not specify a schema

i would suggest "...source does not specify a namespace" instead of "schema". we are trying to add namespace as a primitive into our system, so we shouldn't be afraid to use it. saying schema sorta assumes that the source is a database.

File: airbyte-integrations/connectors/destination-redshift/src/test-integration/java/io/airbyte/integrations/destination/redshift/RedshiftCopyIntegrationTest.java [hunk 12]
```diff
@@ -112,8 +111,9 @@ protected boolean implementsBasicNormalization() {
     return result;
   }
 
-  private List<JsonNode> retrieveRecordsFromTable(String tableName) throws SQLException {
-    final String schemaName = config.get("schema").asText();
+  private List<JsonNode> retrieveRecordsFromTable(String tableName, String schemaName) throws SQLException {
+    System.out.println("Retrieving record:");
```
Comment (cgardens) [inline airbyte-integrations/connectors/destination-redshift/src/test-integration/java/io/airbyte/integrations/destination/redshift/RedshiftCopyIntegrationTest.java:115]: remove or use logger.

File: airbyte-webapp/src/locales/en.json (hunk 11 above)
Comment (cgardens) [reply airbyte-webapp/src/locales/en.json:79]: nice!

File: airbyte-integrations/connectors/destination-bigquery/src/main/resources/spec.json (hunk 7 above)
Comment (cgardens) [reply airbyte-integrations/connectors/destination-bigquery/src/main/resources/spec.json:19]: but reading the actual json, i think it's reasonable.

File: airbyte-integrations/bases/standard-destination-test/src/main/java/io/airbyte/integrations/standardtest/destination/TestDestination.java (hunk 5 above)
Comment (cgardens) [reply airbyte-integrations/bases/standard-destination-test/src/main/java/io/airbyte/integrations/standardtest/destination/TestDestination.java:122]: the diff on this file is hard to parse. i'm assuming it is just 1. a rearrange, 2 adding the defaultNamespace method, and 3 adding this test:  source does not specify a schema. lmk if there's another change in here i should be looking for.

**Review by cgardens — APPROVED** [review]
Very happy with this. Nice work! I especially like how you added support for testing the namespace stuff to testing. It's concise but I think will give us a lot of security. I also appreciate the screenshots you put in to show how your changes will affect UX.

I think one thing that needs to happen (either in this PR or in a follow up) is adding to the docs for each of these destinations what the expect behavior is around namespaces. Maybe just a new subsection we put in the docs for each destination. e.g. the csv, meiliseach, json ones ignore it. BQ should say namespace corresponds to dataset id, etc. I just want to make sure we don't forget it because it's going to be a nuanced idea for our users to wrap their head around. Also you should think about how how we want to document the "namespace" concept in general in our docs. maybe it gets its own readme page in the architecture section alongside incremental.md, etc?

File: airbyte-integrations/bases/standard-destination-test/src/main/java/io/airbyte/integrations/standardtest/destination/TestDestination.java (hunk 5 above)
Comment (davinchia) [reply airbyte-integrations/bases/standard-destination-test/src/main/java/io/airbyte/integrations/standardtest/destination/TestDestination.java:122]: aye apologies. I should have broken the refactoring changes into another commit. Will be better next time.

yes you were looked at all the right bits!

File: airbyte-integrations/connectors/destination-bigquery/src/main/resources/spec.json (hunk 7 above)
Comment (davinchia) [reply airbyte-integrations/connectors/destination-bigquery/src/main/resources/spec.json:19]: oops. got it confused hehe

File: airbyte-integrations/connectors/destination-postgres/src/main/resources/spec.json (hunk 8 above)
Comment (davinchia) [reply airbyte-integrations/connectors/destination-postgres/src/main/resources/spec.json:35]: good point. making the change.

File: airbyte-integrations/connectors/destination-bigquery/src/main/resources/spec.json (hunk 7 above)
Comment (davinchia) [reply airbyte-integrations/connectors/destination-bigquery/src/main/resources/spec.json]: for posterity, this is what this looks like (with the namespace change)

image

File: airbyte-integrations/connectors/destination-redshift/src/test-integration/java/io/airbyte/integrations/destination/redshift/RedshiftCopyIntegrationTest.java (hunk 12 above)
Comment (davinchia) [reply airbyte-integrations/connectors/destination-redshift/src/test-integration/java/io/airbyte/integrations/destination/redshift/RedshiftCopyIntegrationTest.java:115]: oops. thanks for catching this.

Comment (davinchia) [reply airbyte-integrations/connectors/destination-redshift/src/test-integration/java/io/airbyte/integrations/destination/redshift/RedshiftCopyIntegrationTest.java]: removed

**Comment by davinchia** [comment]
Glad this made sense! Good call on documentation @cgardens. Agree with ya. Will release an actual version + add the necessary documentation in a follow up PR.

I was thinking we should release the new destinations + documentation after the source side changes are done (should be in the next few days) since this isn't useful as is yet. What do you think?
